src/Estimators/StructureFactorEstimator.cpp

```diff
@@ -0,0 +1,128 @@
+  num_kpoints_    = pset_ions.getSimulationCell().getKLists().getNumK();
+  kshell_offsets_ = pset_ions.getSimulationCell().getKLists().getKShell();
+  int max_kshell  = kshell_offsets_.size() - 1;
+
+  rhok_tot_r_.resize(num_kpoints_);
+  rhok_tot_i_.resize(num_kpoints_);
+  sfk_e_e_.resize(num_kpoints_);
+  rhok_e_.resize(num_kpoints_);
+  // Legacy comment, but I don't trust it.
+  //for values, we are including e-e structure factor, and e-Ion.  So a total of NumIonSpecies+1 structure factors.
+  //+2 for the real and imaginary parts of rho_k^e
+  //
+  // skAll seems to be written for e-e sf + complex rho_k^e
+  data_.resize(3 * num_kpoints_);
+  kmags_.resize(max_kshell);
+  one_over_degeneracy_kshell_.resize(max_kshell);
+  for (int ks = 0; ks < max_kshell; ks++)
+  {
+    kmags_[ks]                      = std::sqrt(pset_elec.getSimulationCell().getKLists().getKSQWorking()[kshell_offsets_[ks]]);
+    one_over_degeneracy_kshell_[ks] = 1.0 / static_cast<Real>(kshell_offsets_[ks + 1] - kshell_offsets_[ks]);
+  };
+}
+
+StructureFactorEstimator::StructureFactorEstimator(const StructureFactorEstimator& sfe, DataLocality dl)
+    : qmcplusplus::StructureFactorEstimator(sfe)
+{
+  data_locality_ = dl;
+}
+
+void StructureFactorEstimator::accumulate(const RefVector<MCPWalker>& walkers,
+                                          const RefVector<ParticleSet>& psets,
+                                          const RefVector<TrialWaveFunction>& wfns,
+                                          const RefVector<QMCHamiltonian>& hams,
+                                          RandomBase<FullPrecReal>& rng)
+{
+  for (int iw = 0; iw < walkers.size(); ++iw)
+  {
+    Real weight = walkers[iw].get().Weight;
+
+    //sum over species
+    std::copy(psets[iw].get().getSK().rhok_r[0], psets[iw].get().getSK().rhok_r[0] + num_kpoints_, rhok_tot_r_.begin());
+    std::copy(psets[iw].get().getSK().rhok_i[0], psets[iw].get().getSK().rhok_i[0] + num_kpoints_, rhok_tot_i_.begin());
+    for (int i = 1; i < elec_num_species_; ++i)
+      accumulate_elements(psets[iw].get().getSK().rhok_r[i], psets[iw].get().getSK().rhok_r[i] + num_kpoints_,
+                          rhok_tot_r_.begin());
+    for (int i = 1; i < elec_num_species_; ++i)
+      accumulate_elements(psets[iw].get().getSK().rhok_i[i], psets[iw].get().getSK().rhok_i[i] + num_kpoints_,
+                          rhok_tot_i_.begin());
+
+    for (int k = 0; k < num_kpoints_; k++)
+    {
+      sfk_e_e_[k] += weight * (rhok_tot_r_[k] * rhok_tot_r_[k] + rhok_tot_i_[k] * rhok_tot_i_[k]);
+      rhok_e_[k] += weight * std::complex<Real>{rhok_tot_r_[k], rhok_tot_i_[k]};
+    }
+  }
+}
+
+void StructureFactorEstimator::registerOperatorEstimator(hdf_archive& file)
+{
+  hdf_path hdf_name{my_name_};
+  hdf_path path_variables = hdf_name / std::string_view("kpoints");
+  file.push(path_variables, true);
+  file.write(const_cast<std::vector<KPt>&>(ions_.getSimulationCell().getKLists().getKptsCartWorking()), "value");
```

Comment:
Could you add a comment about this `const_cast` to have a piece of mind? Probably a restriction from `write` function?

Reply:
restriction from `hdf_archive.` That are const_cast all over the code like this but you normally see them in calls to observable helper.  I'd like to delete observable helper which i think is an unnecessary state machine so I don't use it here.